apps/omg/lib/omg/validators/merge_transaction_validator.ex

```diff
@@ -0,0 +1,67 @@
+# Copyright 2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.MergeTransactionValidator do
+  @moduledoc """
+  Checks for transaction attributes. Used by OMG.Fees.
+  """
+
+  alias OMG.State.Transaction
+  alias OMG.Utxo
+
+  require Utxo
+
+  use OMG.Utils.LoggerExt
+
+  def is_merge_transaction?(recovered_transaction) do
+    [
+      &is_payment?/1,
+      &has_less_outputs_than_inputs?/1,
+      &has_single_currency?/1,
+      &has_same_account?/1
+    ]
+    |> Enum.all?(fn predicate -> predicate.(recovered_transaction) end)
+  end
+
+  defp is_payment?(%Transaction.Recovered{signed_tx: %{raw_tx: %Transaction.Payment{}}}), do: true
+  defp is_payment?(_), do: false
+
+  defp has_same_account?(%Transaction.Recovered{witnesses: witnesses} = tx) do
+    spenders = Map.values(witnesses)
+
+    tx
+    |> Transaction.get_outputs()
+    |> Enum.map(& &1.owner)
+    |> Enum.concat(spenders)
+    |> single?()
+  end
+
+  defp has_single_currency?(tx) do
+    tx
+    |> Transaction.get_outputs()
+    |> Enum.map(& &1.currency)
+    |> single?()
+  end
+
+  defp has_less_outputs_than_inputs?(tx) do
+    has_less_outputs_than_inputs?(
+      Transaction.get_inputs(tx),
+      Transaction.get_outputs(tx)
+    )
+  end
+
+  defp has_less_outputs_than_inputs?(inputs, outputs), do: length(inputs) >= 1 and length(inputs) > length(outputs)
+
+  defp single?(list), do: 1 == list |> Enum.dedup() |> length()
```

Comment:
I wonder why do we need the `dedup` call?

Reply:
Not sure actually, I didn't really look deep into the actual checking logic, but I'll take a closer look.

Reply:
`single?` is used to check whether whether stuffs are the same in the transaction, e.g. all outputs mention the same recipient, same currency. Actually, I think it is `uniq`, not `dedup`. I wasn't aware `dedup` only removes consecutive duplicates.